File: examples/python-bridges/pythonnet/rpyc_client.py
#!/usr/bin/env python3
"""
Python.NET + RPyC Example Client

Demonstrates using Python.NET to embed CPython in .NET and connect to RPyC.

Prerequisites:
    - .NET Core SDK 6.0+ (or .NET 8 recommended)
    - pip install pythonnet rpyc

Usage:
    1. Start RPyC server: python examples/rpyc-integration/rpyc_server.py
    2. Run this script: python examples/python-bridges/pythonnet/rpyc_client.py

Runtime Configuration:
    Python.NET defaults to .NET Core on modern systems.
    Set PYTHONNET_RUNTIME=coreclr explicitly if needed.
"""

import os
import sys


def configure_runtime():
    """Configure Python.NET to use .NET Core (not Mono)."""
    # Explicitly prefer .NET Core over Mono
    os.environ.setdefault("PYTHONNET_RUNTIME", "coreclr")

    # Optional: specify .NET version
    # os.environ["PYTHONNET_DOTNET_VERSION"] = "8.0.0"


def main():
    # Configure runtime before importing clr
    configure_runtime()

    try:
        import clr
        print("Python.NET loaded successfully")
        print(f"Runtime: {os.environ.get('PYTHONNET_RUNTIME', 'default')}")
    except RuntimeError as e:
        print(f"Error loading Python.NET: {e}")
        print("\nEnsure .NET SDK is installed:")
        print("  Ubuntu/Debian: sudo apt install dotnet-sdk-8.0")
        print("  Or download from: https://dotnet.microsoft.com/download")
        sys.exit(1)

    # Import .NET assemblies
    clr.AddReference("System")
    from System import Console, Environment

    print(f".NET Version: {Environment.Version}")
    print(f"OS: {Environment.OSVersion}")

    # Connect to RPyC server
    import rpyc

    print("\nConnecting to RPyC server on localhost:18812...")
    conn = rpyc.classic.connect("localhost", 18812)

    try:
        # Test 1: Basic math
        math = conn.modules.math
        sqrt_result = math.sqrt(16)
        print(f"math.sqrt(16) = {sqrt_result}")

        # Test 2: NumPy (if available)
        try:
            np = conn.modules.numpy
            arr = np.array([1, 2, 3, 4, 5])
            mean = np.mean(arr)
            print(f"numpy.mean([1,2,3,4,5]) = {mean}")
        except Exception as e:
            print(f"NumPy not available: {e}")

        # Test 3: Get server info (if service provides it)
        try:
            info = conn.root.get_info()
            print(f"Server Python: {info['python_version']}")
        except AttributeError:
            # Classic server doesn't have get_info
            sys_mod = conn.modules.sys
            print(f"Server Python: {sys_mod.version.split()[0]}")

        print("\nAll tests passed!")

    finally:
        conn.close()
        print("Connection closed")


if __name__ == "__main__":
    main()

File: examples/python-bridges/pythonnet/RPyCClient.cs
/**
 * Python.NET + RPyC Example - C# Client
 *
 * Demonstrates calling RPyC from C# via Python.NET embedding.
 *
 * Build and run:
 *   cd examples/python-bridges/pythonnet
 *   dotnet run
 *
 * Prerequisites:
 *   - .NET SDK 6.0+ (8.0 recommended)
 *   - Python 3.8+ with rpyc installed
 */
using System;
using Python.Runtime;

namespace UnifyWeaver.Bridges.PythonNet
{
    class RPyCClient
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Python.NET + RPyC C# Client");
            Console.WriteLine("===========================");

            // Initialize Python runtime
            Runtime.PythonDLL = GetPythonDLL();
            PythonEngine.Initialize();

            try
            {
                using (Py.GIL())
                {
                    // Import RPyC
                    dynamic rpyc = Py.Import("rpyc");

                    // Connect to server
                    Console.WriteLine("\nConnecting to RPyC server...");
                    dynamic conn = rpyc.classic.connect("localhost", 18812);

                    try
                    {
                        // Use remote math module
                        dynamic math = conn.modules.math;
                        double result = math.sqrt(16);
                        Console.WriteLine($"math.sqrt(16) = {result}");

                        // Try NumPy
                        try
                        {
                            dynamic np = conn.modules.numpy;
                            dynamic arr = np.array(new[] { 1, 2, 3, 4, 5 });
                            double mean = np.mean(arr);
                            Console.WriteLine($"numpy.mean([1,2,3,4,5]) = {mean}");
                        }
                        catch (Exception e)
                        {
                            Console.WriteLine($"NumPy not available: {e.Message}");
                        }

                        // Get server info
                        dynamic info = conn.root.get_info();
                        Console.WriteLine($"Server Python: {info["python_version"]}");

                        Console.WriteLine("\nAll tests passed!");
                    }
                    finally
                    {
                        conn.close();
                        Console.WriteLine("Connection closed");
                    }
                }
            }
            finally
            {
                PythonEngine.Shutdown();
            }
        }

        static string GetPythonDLL()
        {
            // Find Python shared library
            // Adjust path based on your Python installation
            if (OperatingSystem.IsLinux())
            {
                return "libpython3.8.so";
            }
            else if (OperatingSystem.IsWindows())
            {
                return "python38.dll";
            }
            else if (OperatingSystem.IsMacOS())
            {
                return "/usr/local/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib";
            }
            throw new PlatformNotSupportedException();
        }
    }
}
